### PR TITLE
make unpublishvolume more robust

### PR DESCRIPTION
**Pull request by TheBeatles1994**

### CODE REVIEW

File: pkg/csi/inflight.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+type InFlight struct {
+	mux      *sync.Mutex
+	inFlight map[string]bool
+}
+
+// NewInFlight instanciates a InFlight structures.
+func NewInFlight() *InFlight {
+	return &InFlight{
+		mux:      &sync.Mutex{},
+		inFlight: make(map[string]bool),
```
Comment (allencloud) [inline pkg/csi/inflight.go:45]: you should use the type of `map[string]struct{}`

File: pkg/csi/nodeserver.go [hunk 2]
```diff
@@ -92,29 +94,37 @@ func newNodeServer(d *csicommon.CSIDriver, dName, nodeID, sysPath string) csi.No
 
 	mounter := k8smount.New("")
 
+	store, err := NewVolumeStore(DefaultEphemeralVolumeDataFilePath)
+	if err != nil {
+		log.Fatal("fail to initialize ephemeral volume store")
```
Comment (allencloud) [inline pkg/csi/nodeserver.go:99]: why not to pass out the detailed err?

File: pkg/csi/volume_store.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+func NewVolumeStore(dataFilePath string) (Store, error) {
+	var err error
+	var volumeDeviceMapper map[string]string
+	if volumeDeviceMapper, err = loadVolumeData(dataFilePath); err != nil {
+		if os.IsNotExist(err) {
+			volumeDeviceMapper = make(map[string]string)
+		} else {
+			return nil, err
+		}
+	}
```
Comment (allencloud) [inline pkg/csi/volume_store.go:46]:
```
volumeDeviceMapper := make(map[string]string)
if _, err = loadVolumeData(dataFilePath); err != nil && !os.IsNotExist(err) {
    return nil, err
}

return &volumeStore{
    volumeDeviceMapper: volumeDeviceMapper,
    dataFilePath:       dataFilePath,
}, nil
```

File: pkg/csi/volume_store.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	return &volumeStore{
+		volumeDeviceMapper: volumeDeviceMapper,
+		dataFilePath:       dataFilePath,
+	}, nil
+}
+
+func (store *volumeStore) AddVolume(volumeID, device string) error {
```
Comment (allencloud) [inline pkg/csi/volume_store.go:53]: do we need to form a rule of what is a valid ID, and the validate it?

File: pkg/csi/inflight.go (hunk 1 above)
Comment (TheBeatles1994) [reply pkg/csi/inflight.go]: done

File: pkg/csi/nodeserver.go (hunk 2 above)
Comment (TheBeatles1994) [reply pkg/csi/nodeserver.go]: done

File: pkg/csi/volume_store.go (hunk 4 above)
Comment (TheBeatles1994) [reply pkg/csi/volume_store.go:53]: Volume id is generated by k8s, so we do not need it.

Comment (allencloud) [reply pkg/csi/volume_store.go:53]: ok

**Pull request merged**